data_mgr.py

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python3
+
+# -*- coding: utf-8 -*-
+# Copyright (C) 2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Create and update the data structure for all workflow services."""
+
+from fnmatch import fnmatchcase
+
+from tornado import gen
+# from google.protobuf.json_format import MessageToDict
+
+# from cylc import flags
+from cylc.exceptions import ClientError, ClientTimeout
+from cylc.ws_messages_pb2 import PbEntireWorkflow
+from cylc.network.scan import MSG_TIMEOUT
+
+
+METHOD_MAP = {
+    'pb_entire_workflow': PbEntireWorkflow
+}
+
+
+async def get_workflow_data(w_id, client, method):
+    # Use already established client
+    try:
+        pb_msg = await client.async_request(method)
+    except ClientTimeout as exc:
+        return (w_id, MSG_TIMEOUT)
+    except ClientError as exc:
+        return (w_id, None)
+    else:
+        ws_data = METHOD_MAP[method]()
+        ws_data.ParseFromString(pb_msg)
+    return (w_id, ws_data)
+
+
+class DataManager(object):
+
+    def __init__(self, ws_mgr):
+        self.ws_mgr = ws_mgr
+        self.data = {}
+
+    # **
+    # Data syncing
+    # **
+    async def entire_workflow_update(self, ids=None):
+        if ids is None:
+            ids = []
+        new_data = {}
+        ws_args = (
+            (w_id, info['req_client'], 'pb_entire_workflow')
+            for w_id, info in self.ws_mgr.workflows.items())
+        items = await gen.multi(
+            get_workflow_data(*args) for args in ws_args
+            if not ids or args[0] in ids)
+        for w_id, result in items:
+            if result is not None and result != MSG_TIMEOUT:
+                new_data[w_id] = result
+        if not ids:
+            # atomic update
+            self.data = new_data
+        else:
+            # apparently faster than update(new_data)
+            for w_id, flow in new_data.items():
+                self.data[w_id] = flow
```

Comment:
Got interested in this one, as I would probably just use `.update(new_data)`. Testing with little data I think `dict.update` was faster?

![image](https://user-images.githubusercontent.com/304786/58054974-7346ef80-7bb0-11e9-82c6-764862eec0a3.png)

Then with a dictionary with some [states and capitals](https://stackoverflow.com/questions/49193164/python-dictionary-states-and-capital-game) the difference for the time running `dict.update` was even bigger.

 
![image](https://user-images.githubusercontent.com/304786/58055136-e486a280-7bb0-11e9-83a8-89b50fe1065f.png)

Even if I initialize the dictionaries with 15 items each, so that the dictionaries are not empty, `dict.update` still appears to be faster.

![image](https://user-images.githubusercontent.com/304786/58055431-dab16f00-7bb1-11e9-93e4-88e2050a0265.png)

Do you have a link to some page explaining why iterating and accessing each item would be faster?

Reply:
Oppsie... I was looking at the speed of the individual action:
https://stackoverflow.com/a/15456218
Yes updating all at once would be better...